src/charm.py

```diff
@@ -37,24 +45,99 @@ def __init__(self, *args):
         self.framework.observe(self.on['grafana-source'].relation_changed,
                                self._on_grafana_changed)
 
-    def _on_config_changed(self, _):
-        """Set a new Juju pod specification
+        if self._stored.provider_ready:
+            self.prometheus_provider = MonitoringProvider(self,
+                                                          'monitoring', 'prometheus', self.version)
+            self.framework.observe(self.prometheus_provider.on.targets_changed,
+                                   self._on_config_changed)
+
+    def _on_config_changed(self, event):
+        """Reconfigure and restart Prometheus.
+
+        In response to any configuration change, such as a new consumer
+        relation, or a new configuration set by the administrator, the
+        Prometheus config file is regenerated, pushed to the workload
+        container and Prometheus is restarted.
+        """
+        logger.info("Handling config changed")
+        container = self.unit.get_container("prometheus")
+
+        # validate configuration options
+        missing_config = self._check_config()
+        if missing_config:
+            logger.error('Incomplete Configuration : {}. '
+                         'Application will be blocked.'.format(missing_config))
+            self.unit.status = \
+                BlockedStatus('Missing configuration: {}'.format(missing_config))
+            return
+
+        # check if configuration file has changed and if so push the
+        # new config file to the workload container
+        prometheus_config = self._prometheus_config()
+        config_hash = str(hashlib.md5(str(prometheus_config).encode('utf-8')))
+        if not self._stored.prometheus_config_hash == config_hash:
+            self._stored.prometheus_config_hash = config_hash
+            container.push(PROMETHEUS_CONFIG, prometheus_config)
+            logger.info("Pushed new configuration")
+
+        # setup the workload (Prometheus) container and its services
+        layer = self._prometheus_layer()
+        plan = container.get_plan()
+        if plan.services != layer["services"]:
+            container.add_layer("prometheus", layer, combine=True)
+
+            if container.get_service("prometheus").is_running():
+                container.stop("prometheus")
+
+            container.start("prometheus")
+            logger.info("Restarted prometheus done")
```

Comment:
Minor nitpick here. This will say restarting even if it is being started for the first time.

Reply:
Will reword in a future commit. Feel free to remind me, if I forget.